Ladybird WebContent Crash

### Summary

We visiting http://www.zed.com and allowing the site to load. When moving the cursor over the menubar or the hero section that highlights the elements, the site crashes.

I removed the JS from the SingleFile and it still crashed, leading me to assume the issue could be CSS related given the animation is what causes the crash.

### Operating system

macOS

### Steps to reproduce

1. Go to http://www.zed.com
2. Once loaded move the cursor over the menu bar/hero section to enable interactive element highlighting


### Expected behavior

Menu bar/Hero section link highliting

### Actual behavior

Website crash.

### URL for a reduced test case

https://gist.github.com/seth-wood/18836d31d028cbb49c2dd4673a06a3a0

### HTML/SVG/etc. source for a reduced test case

```html
N/A
```


### Log output and (if possible) backtrace


```shell
46514.146 WebContent(19406): (js log) "JQMIGRATE: Migrate is installed, version 3.4.0"
46514.161 WebContent(19406): No property (from 4 properties) matched Token: Comma
46514.161 WebContent(19406):     animation-direction
46514.161 WebContent(19406):     animation-play-state
46514.161 WebContent(19406):     animation-delay
46514.161 WebContent(19406):     animation-fill-mode
46514.161 WebContent(19406): No property (from 4 properties) matched Token: Comma
46514.161 WebContent(19406):     animation-direction
46514.161 WebContent(19406):     animation-play-state
46514.161 WebContent(19406):     animation-delay
46514.161 WebContent(19406):     animation-fill-mode
46514.485 WebContent(19406): ResourceLoader: Filtered request to: "https://www.googletagmanager.com/gtag/js?id=UA-114649197-1"
46514.495 WebContent(19406): HTMLScriptElement: Refusing to run script because the element's result is null.
46514.650 WebContent(19406): ResourceLoader: Failed load of: "https://www.zed.com/wp-includes/js/font.js", Error: Load failed (status: 404 Not Found), Duration: 165ms
46514.650 WebContent(19406): HTMLScriptElement: Refusing to run script because the element's result is null.
46514.656 WebContent(19406): ResourceLoader: Filtered request to: "https://connect.facebook.net/en_US/sdk.js#xfbml=1&version=v2.5"
46514.658 WebContent(19406): HTMLScriptElement: Refusing to run script because the element's result is null.
46515.798 Ladybird(19301): WebContent process crashed! Last page loaded: https://www.zed.com/
46515.798 Ladybird(19301): Consider raising an issue at https://github.com/LadybirdBrowser/ladybird/issues/new/choose

*****************************************


Local Reduction.html log (with JS removed):
GrMtlRenderPipeline: 0x157138800 (4 refs)
GrMtlDepthStencil: 0x6000011c77f0 (2 refs)
GrMtlRenderPipeline: 0x16082ba00 (3 refs)
GrMtlSampler: 0x6000005dc2c0 (1 refs)
GrMtlRenderPipeline: 0x160839400 (3 refs)
../src/335264b559-747e4e3c5f.clean/src/gpu/ganesh/GrManagedResource.h:61: fatal error: "check(0 == fHashSet.count())"
47510.664 WebContent(19573): FIXME: Implement 'content-security-policy' http-equiv state
47518.521 Ladybird(19301): WebContent process crashed! Last page loaded: file:///Users/sethwood/Downloads/Reduction.html
47518.521 Ladybird(19301): Consider raising an issue at https://github.com/LadybirdBrowser/ladybird/issues/new/choose
```


### Screenshots or screen recordings

_No response_

### Build flags or config settings

_No response_

### Contribute a patch?

- [ ] I’ll contribute a patch for this myself.

--- a/Libraries/LibWeb/Page/EventHandler.cpp
+++ b/Libraries/LibWeb/Page/EventHandler.cpp
@@ -148,7 +148,7 @@ static Gfx::StandardCursor cursor_css_to_gfx(Optional<CSS::Cursor> cursor)
 }
 
 // https://drafts.csswg.org/cssom-view/#dom-mouseevent-offsetx
-static CSSPixelPoint compute_mouse_event_offset(CSSPixelPoint position, Layout::Node const& layout_node)
+static CSSPixelPoint compute_mouse_event_offset(CSSPixelPoint position, Painting::Paintable const& paintable)
 {
     // If the event’s dispatch flag is set,
     // FIXME: Is this guaranteed to be dispatched?
@@ -160,11 +160,11 @@ static CSSPixelPoint compute_mouse_event_offset(CSSPixelPoint position, Layout::
     };
 
     // ignoring the transforms that apply to the element and its ancestors,
-    if (layout_node.has_css_transform()) {
-        auto const& paintable_box = layout_node.dom_node()->paintable_box();
-        auto const affine_transform = Gfx::extract_2d_affine_transform(paintable_box->transform().inverse());
+    if (paintable.layout_node().has_css_transform()) {
+        auto const& paintable_box = static_cast<Painting::PaintableBox const&>(paintable);
+        auto const affine_transform = Gfx::extract_2d_affine_transform(paintable_box.transform().inverse());
 
-        auto const& origin = paintable_box->transform_origin();
+        auto const& origin = paintable_box.transform_origin();
         Gfx::Point<float> const precision_origin = {
             origin.x().to_double(),
             origin.y().to_double()
@@ -176,7 +176,7 @@ static CSSPixelPoint compute_mouse_event_offset(CSSPixelPoint position, Layout::
     }
 
     // relative to the origin of the padding edge of the target node
-    auto const top_left_of_layout_node = layout_node.first_paintable()->box_type_agnostic_position();
+    auto const top_left_of_layout_node = paintable.box_type_agnostic_position();
     CSSPixelPoint offset = {
         CSSPixels(precision_offset.x()),
         CSSPixels(precision_offset.y())
@@ -252,7 +252,7 @@ EventResult EventHandler::handle_mousewheel(CSSPixelPoint viewport_position, CSS
             // FIXME: Support wheel events in nested browsing contexts.
             if (is<HTML::HTMLIFrameElement>(*node)) {
                 auto& iframe = static_cast<HTML::HTMLIFrameElement&>(*node);
-                auto position_in_iframe = viewport_position.translated(compute_mouse_event_offset({}, paintable->layout_node()));
+                auto position_in_iframe = viewport_position.translated(compute_mouse_event_offset({}, *paintable));
                 iframe.content_navigable()->event_handler().handle_mousewheel(position_in_iframe, screen_position, button, buttons, modifiers, wheel_delta_x, wheel_delta_y);
                 return EventResult::Dropped;
             }
@@ -263,7 +263,7 @@ EventResult EventHandler::handle_mousewheel(CSSPixelPoint viewport_position, CSS
                 return EventResult::Dropped;
 
             auto page_offset = compute_mouse_event_page_offset(viewport_position);
-            auto offset = compute_mouse_event_offset(page_offset, *layout_node);
+            auto offset = compute_mouse_event_offset(page_offset, *layout_node->first_paintable());
             if (node->dispatch_event(UIEvents::WheelEvent::create_from_platform_event(node->realm(), UIEvents::EventNames::wheel, screen_position, page_offset, viewport_position, offset, wheel_delta_x, wheel_delta_y, button, buttons, modifiers).release_value_but_fixme_should_propagate_errors())) {
                 m_navigable->active_window()->scroll_by(wheel_delta_x, wheel_delta_y);
             }
@@ -314,7 +314,7 @@ EventResult EventHandler::handle_mouseup(CSSPixelPoint viewport_position, CSSPix
         if (node) {
             if (is<HTML::HTMLIFrameElement>(*node)) {
                 if (auto content_navigable = static_cast<HTML::HTMLIFrameElement&>(*node).content_navigable())
-                    return content_navigable->event_handler().handle_mouseup(viewport_position.translated(compute_mouse_event_offset({}, paintable->layout_node())), screen_position, button, buttons, modifiers);
+                    return content_navigable->event_handler().handle_mouseup(viewport_position.translated(compute_mouse_event_offset({}, *paintable)), screen_position, button, buttons, modifiers);
                 return EventResult::Dropped;
             }
 
@@ -328,7 +328,7 @@ EventResult EventHandler::handle_mouseup(CSSPixelPoint viewport_position, CSSPix
             }
 
             auto page_offset = compute_mouse_event_page_offset(viewport_position);
-            auto offset = compute_mouse_event_offset(page_offset, *layout_node);
+            auto offset = compute_mouse_event_offset(page_offset, *layout_node->first_paintable());
             node->dispatch_event(UIEvents::PointerEvent::create_from_platform_event(node->realm(), UIEvents::EventNames::pointerup, screen_position, page_offset, viewport_position, offset, {}, button, buttons, modifiers).release_value_but_fixme_should_propagate_errors());
             node->dispatch_event(UIEvents::MouseEvent::create_from_platform_event(node->realm(), UIEvents::EventNames::mouseup, screen_position, page_offset, viewport_position, offset, {}, button, buttons, modifiers).release_value_but_fixme_should_propagate_errors());
             handled_event = EventResult::Handled;
@@ -453,7 +453,7 @@ EventResult EventHandler::handle_mousedown(CSSPixelPoint viewport_position, CSSP
 
         if (is<HTML::HTMLIFrameElement>(*node)) {
             if (auto content_navigable = static_cast<HTML::HTMLIFrameElement&>(*node).content_navigable())
-                return content_navigable->event_handler().handle_mousedown(viewport_position.translated(compute_mouse_event_offset({}, paintable->layout_node())), screen_position, button, buttons, modifiers);
+                return content_navigable->event_handler().handle_mousedown(viewport_position.translated(compute_mouse_event_offset({}, *paintable)), screen_position, button, buttons, modifiers);
             return EventResult::Dropped;
         }
 
@@ -468,7 +468,7 @@ EventResult EventHandler::handle_mousedown(CSSPixelPoint viewport_position, CSSP
 
         m_mousedown_target = node.ptr();
         auto page_offset = compute_mouse_event_page_offset(viewport_position);
-        auto offset = compute_mouse_event_offset(page_offset, *layout_node);
+        auto offset = compute_mouse_event_offset(page_offset, *layout_node->first_paintable());
         node->dispatch_event(UIEvents::PointerEvent::create_from_platform_event(node->realm(), UIEvents::EventNames::pointerdown, screen_position, page_offset, viewport_position, offset, {}, button, buttons, modifiers).release_value_but_fixme_should_propagate_errors());
         node->dispatch_event(UIEvents::MouseEvent::create_from_platform_event(node->realm(), UIEvents::EventNames::mousedown, screen_position, page_offset, viewport_position, offset, {}, button, buttons, modifiers).release_value_but_fixme_should_propagate_errors());
     }
@@ -573,7 +573,7 @@ EventResult EventHandler::handle_mousemove(CSSPixelPoint viewport_position, CSSP
 
         if (node && is<HTML::HTMLIFrameElement>(*node)) {
             if (auto content_navigable = static_cast<HTML::HTMLIFrameElement&>(*node).content_navigable())
-                return content_navigable->event_handler().handle_mousemove(viewport_position.translated(compute_mouse_event_offset({}, paintable->layout_node())), screen_position, buttons, modifiers);
+                return content_navigable->event_handler().handle_mousemove(viewport_position.translated(compute_mouse_event_offset({}, *paintable)), screen_position, buttons, modifiers);
             return EventResult::Dropped;
         }
 
@@ -607,7 +607,7 @@ EventResult EventHandler::handle_mousemove(CSSPixelPoint viewport_position, CSSP
             }
 
             auto page_offset = compute_mouse_event_page_offset(viewport_position);
-            auto offset = compute_mouse_event_offset(page_offset, *layout_node);
+            auto offset = compute_mouse_event_offset(page_offset, *layout_node->first_paintable());
             auto movement = compute_mouse_event_movement(screen_position);
 
             m_mousemove_previous_screen_position = screen_position;
@@ -716,7 +716,7 @@ EventResult EventHandler::handle_doubleclick(CSSPixelPoint viewport_position, CS
 
     if (is<HTML::HTMLIFrameElement>(*node)) {
         if (auto content_navigable = static_cast<HTML::HTMLIFrameElement&>(*node).content_navigable())
-            return content_navigable->event_handler().handle_doubleclick(viewport_position.translated(compute_mouse_event_offset({}, paintable->layout_node())), screen_position, button, buttons, modifiers);
+            return content_navigable->event_handler().handle_doubleclick(viewport_position.translated(compute_mouse_event_offset({}, *paintable)), screen_position, button, buttons, modifiers);
         return EventResult::Dropped;
     }
 
@@ -727,7 +727,7 @@ EventResult EventHandler::handle_doubleclick(CSSPixelPoint viewport_position, CS
         return EventResult::Dropped;
 
     auto page_offset = compute_mouse_event_page_offset(viewport_position);
-    auto offset = compute_mouse_event_offset(page_offset, *layout_node);
+    auto offset = compute_mouse_event_offset(page_offset, *layout_node->first_paintable());
     node->dispatch_event(UIEvents::MouseEvent::create_from_platform_event(node->realm(), UIEvents::EventNames::dblclick, screen_position, page_offset, viewport_position, offset, {}, button, buttons, modifiers).release_value_but_fixme_should_propagate_errors());
 
     // NOTE: Dispatching an event may have disturbed the world.
@@ -784,12 +784,12 @@ EventResult EventHandler::handle_drag_and_drop_event(DragEvent::Type type, CSSPi
 
     if (is<HTML::HTMLIFrameElement>(*node)) {
         if (auto content_navigable = static_cast<HTML::HTMLIFrameElement&>(*node).content_navigable())
-            return content_navigable->event_handler().handle_drag_and_drop_event(type, viewport_position.translated(compute_mouse_event_offset({}, paintable->layout_node())), screen_position, button, buttons, modifiers, move(files));
+            return content_navigable->event_handler().handle_drag_and_drop_event(type, viewport_position.translated(compute_mouse_event_offset({}, *paintable)), screen_position, button, buttons, modifiers, move(files));
         return EventResult::Dropped;
     }
 
     auto page_offset = compute_mouse_event_page_offset(viewport_position);
-    auto offset = compute_mouse_event_offset(page_offset, paintable->layout_node());
+    auto offset = compute_mouse_event_offset(page_offset, *paintable);
 
     switch (type) {
     case DragEvent::Type::DragStart:

--- a/Tests/LibWeb/Text/expected/CSSOMView/mouse-event-transformed-pseudoelement-crash.txt
+++ b/Tests/LibWeb/Text/expected/CSSOMView/mouse-event-transformed-pseudoelement-crash.txt
@@ -1,0 +1,1 @@
+PASS (didn't crash)

--- a/Tests/LibWeb/Text/input/CSSOMView/mouse-event-transformed-pseudoelement-crash.html
+++ b/Tests/LibWeb/Text/input/CSSOMView/mouse-event-transformed-pseudoelement-crash.html
@@ -1,0 +1,21 @@
+<!DOCTYPE html>
+<style>
+    div::before {
+        content: "";
+        display: block;
+        width: 100px;
+        height: 100px;
+        background-color: gray;
+        transform: rotate(45deg);
+    }
+</style>
+<div></div>
+
+<script src="../include.js"></script>
+
+<script>
+    test(() => {
+        internals.click(50, 60);
+        println("PASS (didn't crash)");
+    });
+</script>

